File: core/contractsdk/cpp/example/hash_deposit.cc
#include "xchain/xchain.h"
#include <string>

const std::string UserBucket = "USER";
const std::string HashBucket = "HASH";

// 文件Hash存证API规范
// 参数由Context提供
class HashDepositBasic {
public:
    // 初始化，基本不做任何工作
    virtual void initialize() = 0;
    // 将用户参数{user_id,file_name,hash_id}存储到磁盘
    virtual void storeFileInfo() = 0;
    // 查询当前合约下所有用户
    virtual void queryUserList() = 0;
    // 查询某个User下所有信息
    virtual void queryFileInfoByUser() = 0;
    // 按照Hash查询文件信息,需要指定用户
    virtual void queryFileInfoByHash() = 0;
};

struct HashDeposit : public HashDepositBasic, public xchain::Contract {
public:
    void initialize() {
        xchain::Context* ctx = this->context();
        ctx->ok("initialize success");
    }
    void storeFileInfo() {
        xchain::Context* ctx = this->context();    
        std::string user_id = ctx->arg("user_id");
        std::string hash_id = ctx->arg("hash_id");
        std::string file_name = ctx->arg("file_name");
        const std::string userKey = UserBucket + "/" + user_id + "/" + hash_id;
        const std::string hashKey = HashBucket + "/" + hash_id;
        std::string value = user_id + "\t" + hash_id + "\t" + file_name;
        std::string tempVal;
        if (ctx->get_object(hashKey, &tempVal)) {
            ctx->error("storeFileInfo failed, such hash has existed already");
            return;
        }
        if (ctx->put_object(userKey, value) && ctx->put_object(hashKey, value)) {
            ctx->ok("storeFileInfo success");
            return;
        }
        ctx->error("storeFileInfo failed");
    }
    
    void queryUserList() {
        xchain::Context* ctx = this->context();
        const std::string key = UserBucket + "/";
        std::unique_ptr<xchain::Iterator> iter = ctx->new_iterator(key, key + "～");
        std::string result;
        while (iter->next()) {
            std::pair<std::string, std::string> res;
            iter->get(&res);    
            if (res.first.length() > UserBucket.length() + 1) {
                result += res.first.substr(UserBucket.length() + 1) + "\n";
            }
        }
        ctx->ok(result);
    }
    void queryFileInfoByUser() {
        xchain::Context* ctx = this->context();
        const std::string key = UserBucket + "/" + ctx->arg("user_id");
        std::unique_ptr<xchain::Iterator> iter = ctx->new_iterator(key, key + "～");
        std::string result;
        while (iter->next()) {
            std::pair<std::string, std::string> res;
            iter->get(&res);
            result += res.second + "\n";
        }
        ctx->ok(result);
    }
    void queryFileInfoByHash() {
        xchain::Context* ctx = this->context();
        
        const std::string key = HashBucket + "/" + ctx->arg("hash_id");
        std::string value;
        bool ret = ctx->get_object(key, &value);
        if (ret) {
            ctx->ok(value);
            return;
        }
        ctx->error("queryFileInfoByHash error");
    }
};

DEFINE_METHOD(HashDeposit, initialize) {
    self.initialize();
}

DEFINE_METHOD(HashDeposit, storeFileInfo) {
    self.storeFileInfo();
}

DEFINE_METHOD(HashDeposit, queryUserList) {
    self.queryUserList();
}

DEFINE_METHOD(HashDeposit, queryFileInfoByUser) {
    self.queryFileInfoByUser();
}

DEFINE_METHOD(HashDeposit, queryFileInfoByHash) {
    self.queryFileInfoByHash();
}

File: core/contractsdk/java/example/hash-deposit/src/main/java/com/baidu/xuper/HashDeposit.java
package com.baidu.xuper;

import java.util.Iterator;

/**
 * Counter
 */
public class HashDeposit implements Contract {
    final String USER_BUCKET = "USER";
    final String HASH_BUCKET = "HASH";

    @Override
    @ContractMethod
    public Response initialize(Context ctx) {
        return Response.ok("ok".getBytes());
    }

    @ContractMethod
    public Response storeFileInfo(Context ctx) {
        if (ctx.args().get("user_id") == null || ctx.args().get("user_id").length == 0 ||
                ctx.args().get("hash_id") == null || ctx.args().get("hash_id").length == 0 ||
                ctx.args().get("file_name") == null || ctx.args().get("file_name").length == 0
        ) {
            return Response.error("missing user_id or hash_id or filename");
        }
        String user_id = new String(ctx.args().get("user_id"));
        String hash_id = new String(ctx.args().get("hash_id"));
        String file_name = new String(ctx.args().get("file_name"));

        String userKey = USER_BUCKET + "/" + user_id + "/" + hash_id;
        String hashKey = HASH_BUCKET + "/" + hash_id;
        String value = user_id + "\t" + hash_id + "\t" + file_name;
        if (ctx.getObject(hashKey.getBytes()) != null) {
            return Response.error("hashid " + hash_id + " already exists");
        }
        ctx.putObject(userKey.getBytes(), value.getBytes());
        ctx.putObject(hashKey.getBytes(), value.getBytes());
        return Response.ok("".getBytes());
    }

    @ContractMethod
    public Response queryUserList(Context ctx) {
        String key = USER_BUCKET + "/";
        String start = key;
        String end = key + "~";
        Iterator<ContractIteratorItem> iter = ctx.newIterator(start.getBytes(), end.getBytes());
        StringBuffer buf = new StringBuffer();
        iter.forEachRemaining(
                item -> {
                    buf.append(new String(item.getKey()).split("/")[1]);
                    buf.append("\t");
                }
        );
        return Response.ok(buf.toString().getBytes());
    }

    @ContractMethod
    public Response queryFileInfoByUser(Context ctx) {
        String key = USER_BUCKET + "/" + new String(ctx.args().get("user_id"));
        String start = key;
        String end = start + "~";
        StringBuffer buf = new StringBuffer();
        ctx.newIterator(start.getBytes(), end.getBytes()).forEachRemaining(
                item -> {
                    buf.append(new String(item.getValue()));
                    buf.append("\n");
                }
        );
        return Response.ok(buf.toString().getBytes());
    }

    @ContractMethod
    public Response queryFileInfoByHash(Context ctx) {
        byte[] hashId = ctx.args().get("hash_id");
        if (hashId == null || hashId.length == 0) {
            return Response.error("missing hash_id");
        }
        String key = HASH_BUCKET + "/" + new String(hashId);
        byte[] info = ctx.getObject(key.getBytes());
        if (info == null) {
            return Response.error("file info not exists");
        }
        return Response.ok(info);
    }

    public static void main(String[] args) {
        Driver.serve(new HashDeposit());
    }
}